main/src/cgeo/geocaching/calculator/FormulaParser.java

```diff
@@ -0,0 +1,379 @@
+        }
+
+        /**
+         * Parses latitude or longitude out of a given range of matched groups.
+         *
+         * @param matcher
+         *            the matcher that holds the matches groups
+         * @param first
+         *            the first group to parse
+         * @param last
+         *            the last group to parse
+         * @return the parsed latitude/longitude, or null if parsing failed
+         */
+        @Nullable
+        private String parseGroups(@NonNull final MatcherWrapper matcher, final int first, final int last) {
+            final List<String> groups = new ArrayList<>(last - first + 1);
+            for (int i = first; i <= last; i++) {
+                groups.add(matcher.group(i));
+            }
+
+            return parse(groups);
+        }
+
+        /**
+         * @see AbstractFormulaParser#parse(String)
+         */
+        @Override
+        @Nullable
+        public final ImmutablePair<String, String> parse(@NonNull final String text) {
+            final MatcherWrapper matcher = new MatcherWrapper(latLonPattern, text);
+            if (matcher.find()) {
+                final int groupCount = matcher.groupCount();
+                final int partCount = groupCount / 2;
+
+                final String lat = parseGroups(matcher, 1, partCount);
+                if (lat == null || lat.isEmpty()) {
+                    return null;
+                }
+
+                final String lon = parseGroups(matcher, partCount + 1, groupCount);
+                if (lon == null || lon.isEmpty()) {
+                    return null;
+                }
+
+                return new ImmutablePair<>(lat, lon);
+            }
+
+            return null;
+        }
+
+        /**
+         * @see AbstractFormulaParser#parse(String, Geopoint.LatLon)
+         */
+        @Override
+        @Nullable
+        public final ResultWrapper parse(@NonNull final String text, @NonNull final Geopoint.LatLon latlon) {
+            final MatcherWrapper matcher = new MatcherWrapper(latlon == Geopoint.LatLon.LAT ? latPattern : lonPattern, text);
+            if (matcher.find()) {
+                final String res = parseGroups(matcher, 1, matcher.groupCount());
+                if (res != null) {
+                    return new ResultWrapper(res, matcher.group().length());
+                }
+            }
+
+            return null;
+        }
+
+        /**
+         * Parses latitude or longitude from matched groups of corresponding pattern.
+         *
+         * @param groups
+         *            the groups matched by latitude/longitude pattern
+         * @return parsed latitude/longitude, or null if parsing failed
+         */
+        @Nullable
+        public abstract String parse(@NonNull List<String> groups);
+    }
+
+
+    /**
+     * Parser for MinDec format: X DD° MM.MMM'.
+     */
+    private static final class MinDecFormulaParser extends AbstractLatLonFormulaParser {
```

Comment:
Is MinDecFormulaParser the only subclass of AbstractLatLonFormulaParser? If yes, why do we need the subclass?

Reply:
No, there can / should be the others as well, but with your next comment, probably that will be done with reusing the other parser...